lib/remap-parser/src/parser.lalrpop

```diff
@@ -0,0 +1,501 @@
+use std::str::FromStr;
+use crate::ast::*;
+use crate::lex::*;
+use diagnostic::span;
+use std::collections::BTreeMap;
+use lalrpop_util::ErrorRecovery;
+use ordered_float::NotNan;
+
+grammar<'err, 'input>(input: &'input str);
+
+extern {
+    type Location = usize;
+    type Error = Error;
+
+    enum Tok<'input> {
+        "identifier" => Token::Identifier(<&'input str>),
+        "string literal" => Token::StringLiteral(<StringLiteral<&'input str>>),
+        "integer literal" => Token::IntegerLiteral(<i64>),
+        "float literal" => Token::FloatLiteral(<NotNan<f64>>),
+        "regex literal" => Token::RegexLiteral(<&'input str>),
+        "timestamp literal" => Token::TimestampLiteral(<&'input str>),
+
+        "function call" => Token::FunctionCall(<&'input str>),
+
+        "invalid token" => Token::InvalidToken(<char>),
+
+        "reserved identifier" => Token::ReservedIdentifier(<&'input str>),
+
+        LQuery => Token::LQuery,
+        RQuery => Token::RQuery,
+
+        "if" => Token::If,
+        "else" => Token::Else,
+        "null" => Token::Null,
+        "true" => Token::True,
+        "false" => Token::False,
+
+        ";" => Token::SemiColon,
+        "\n" => Token::Newline,
+        "?" => Token::Question,
+        "|" => Token::Pipe,
+        "=" => Token::Equals,
+        "," => Token::Comma,
+        "_" => Token::Underscore,
+        ":" => Token::Colon,
+        "." => Token::Dot,
+        "!" => Token::Bang,
+
+        "+" => Token::Operator("+"),
+        "*" => Token::Operator("*"),
+        "-" => Token::Operator("-"),
+        "/" => Token::Operator("/"),
+        "%" => Token::Operator("%"),
+        "<" => Token::Operator("<"),
+        "<=" => Token::Operator("<="),
+        ">" => Token::Operator(">"),
+        ">=" => Token::Operator(">="),
+        "==" => Token::Operator("=="),
+        "!=" => Token::Operator("!="),
+
+        "&&" => Token::Operator("&&"),
+        "||" => Token::Operator("||"),
+        "??" => Token::Operator("??"),
+
+        "[" => Token::LBracket,
+        "{" => Token::LBrace,
+        "(" => Token::LParen,
+
+        "]" => Token::RBracket,
+        "}" => Token::RBrace,
+        ")" => Token::RParen,
+
+        "_r" => Token::InternalTest("_r"),
+        "_e" => Token::InternalTest("_e"),
+        "_m" => Token::InternalTest("_m"),
+        "_a" => Token::InternalTest("_a"),
+        "_q" => Token::InternalTest("_q"),
+        "_c" => Token::InternalTest("_c"),
+        "_as" => Token::InternalTest("_as"),
+        "_fn" => Token::InternalTest("_fn"),
+    }
+}
+
+// -----------------------------------------------------------------------------
+// parser entry points
+// -----------------------------------------------------------------------------
+
+// The main entrypoint into a VRL program.
+//
+// A program consists of one or more expressions.
+pub Program: Program = NonterminalNewline* <RootExprs> => Program(<>);
```

Comment:
Quick guide on how to read this syntax:

```rust
 pub Program: Program = NonterminalNewline* <RootExprs> => Program(<>);
```

* `pub` — make this rule public, basically generating a `<Rule>Parser::parse()` function (e.g. `ProgramParser`)
* `Program`  — the name of the non-terminal
* `: Program` — the (Rust) type returned by this non-terminal
* `= NonterminalNewline* <RootExprs>` — the matching terminal (e.g. a sequence of lexed tokens that should match for this non-terminal to match)
	* furthermore, the `<...>` indicates that only `RootExprs` is provided to the _action_ of the match arm
* `=>` start the _action_ of the match arm, anything after this token is regular Rust code*.
	* the exception being `<>`, which is short for "whatever is captured by the match arm".

For more details, see: https://lalrpop.github.io/lalrpop/tutorial/003_type_inference.html